notebooks/02-ingestion/05-partitioning-data.ipynb

```diff
@@ -0,0 +1,717 @@
+    "  \"Div4LongestGTime\",  \"Div4WheelsOff\",  \"Div4TailNum\",  \"Div5Airport\",  \"Div5AirportID\",  \"Div5AirportSeqID\",  \"Div5WheelsOn\",  \"Div5TotalGTime\",\n",
+    "  \"Div5LongestGTime\",  \"Div5WheelsOff\",  \"Div5TailNum\"\n",
+    "FROM \"ext\"\n",
+    "PARTITIONED BY YEAR\n",
+    "CLUSTERED BY IATA_CODE_Reporting_Airline\n",
+    "'''\n",
+    "# use a request so that we can specify the query context\n",
+    "req = sql_client.sql_request(sql)\n",
+    "req.add_context(\"rowsPerSegment\", \"50000\")  # we still want to target our ideal of 50k per segment, normally greater than 3 million \n",
+    "display.run_task(req)\n",
+    "sql_client.wait_until_ready('flights_year_IATA')"
+   ]
+  },
+  {
+   "attachments": {},
+   "cell_type": "markdown",
+   "id": "8799d234-c2f3-4277-9661-8931b94d58ff",
+   "metadata": {},
+   "source": [
+    "The [segment view for this table](http://localhost:8888/unified-console.html#segments/datasource~flights_year_IATA) shows 12 segments that are still very close to our ideal segment size, but now they have some new metadata based on the clustering columns in the `Shard Spec` column. It shows which range of values of the IATA_CODE_Reporting_Airline column are available in the segment. This is very useful for pruning when filtering on that column because Druid can prune to 1 or 2 segment files when looking for a single airline.\n",
+    "\n",
+    "![](assets/segments-iata.png)\n",
+    "\n",
+    "## Query Performance\n",
+    "Now that the data has been loaded using a few different partitioning strategies, you can use a test query to determine which one is the best.\n",
+    "In order to control the experiment we will turn cacheing off for the results of the query and measure the same query with each table.\n",
+    "\n",
+    "### The Query\n",
+    "It is well known that flight traffic follows a weekly pattern as well as a seasonal pattern. Given that we only have a month, let's take a look at the maximum and average delay for a single airline by day of the week so that we can see the pattern and how it affects delays. We'll only consider the departure delay for now and only use delays greater than 2 minutes as \"delayed\".\n",
+    "\n",
+    "One caveat about measuring performance on a laptop. The results obtained here will not necessarily be the best when the data grows and is distributed on a multi-node cluster. The intent here is to provide a way of thinking about optimizing the data. In a real cluster, it is likely that the results will vary and a different partitioning and clustering strategy will provide better results.\n",
+    "\n",
+    "The following cell defines a function we can use to measure performance:"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "77a71e6a-73a2-4bd7-9e8d-fd7cb966da65",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "from datetime import datetime\n",
```

Comment:
imports should be at the top of the notebook

Reply:
agreed...done